src/typescript/typescript-document.ts

```diff
@@ -0,0 +1,56 @@
+/**
+ * @license
+ * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import * as ts from 'typescript';
+import {Node, Program} from 'typescript';
+
+import {SourceRange} from '../model/model';
+import {Options, ParsedDocument, StringifyOptions} from '../parser/document';
+
+import {Visitor} from './typescript-visitor';
+
+export {Node, Program} from 'typescript';
+export {Options} from '../parser/document';
+export {Visitor} from './typescript-visitor';
+
+export interface TSOptions extends Options<Node> { program: Program; }
+
+export class ParsedTypeScriptDocument extends ParsedDocument<Node, Visitor> {
+  type = 'typescript';
+  program: Program;
+
+  constructor(from: TSOptions) {
+    super(from);
+    this.program = from.program;
+  }
+
+  visit(visitors: Visitor[]) {
+    const sourceFile = this.program.getSourceFile(this.url);
+    for (const visitor of visitors) {
+      ts.forEachChild(sourceFile, (node) => visitor.visitNode(node));
+    }
+  }
+
+  forEachNode(_callback: (node: Node) => void) {
```

Comment:
Feel free to punt, but I think we can drop this as a required method on ParsedDocument.